src/cascadia/TerminalApp/VsSetupConfiguration.h

```diff
@@ -0,0 +1,142 @@
+        typedef wil::com_ptr<ISetupInstanceCatalog> ComPtrInstanceCatalog;
+
+    public:
+        struct VsSetupInstance
+        {
+            inline const static std::wstring_view kVersion16_3{ L"16.3.0.0" };
+
+            inline std::wstring ResolvePath(std::wstring_view relativePath) const
+            {
+                return VsSetupConfiguration::ResolvePath(mp_inst, relativePath);
+            }
+
+            inline std::wstring GetDevShellModulePath() const
+            {
+                unsigned long long ullInstanceVersion{ 0 };
+                THROW_IF_FAILED(mp_helper->ParseVersion(GetVersion().data(), &ullInstanceVersion));
+
+                unsigned long long ullVersion16_3{ 0 };
+                THROW_IF_FAILED(mp_helper->ParseVersion(kVersion16_3.data(), &ullVersion16_3));
+
+                if (ullInstanceVersion >= ullVersion16_3)
+                    return ResolvePath(L"Common7\\Tools\\Microsoft.VisualStudio.DevShell.dll");
+
+                return ResolvePath(L"Common7\\Tools\\vsdevshell\\Microsoft.VisualStudio.DevShell.dll");
+            }
+
+            inline std::wstring GetDevCmdScriptPath() const
+            {
+                return ResolvePath(L"Common7\\Tools\\VsDevCmd.bat");
+            }
+
+            inline bool VersionInRange(std::wstring_view range) const
+            {
+                return VsSetupConfiguration::InstallationVersionInRange(mp_query, mp_inst, range);
+            }
+
+            inline std::wstring GetVersion() const
+            {
+                return VsSetupConfiguration::GetInstallationVersion(mp_inst);
+            }
+
+            inline std::wstring GetInstallationPath() const
+            {
+                return VsSetupConfiguration::GetInstallationPath(mp_inst);
+            }
+
+            inline std::wstring GetInstanceId() const
+            {
+                return VsSetupConfiguration::GetInstanceId(mp_inst);
+            }
+
+            inline std::wstring GetChannelId() const
+            {
+                ComPtrPropertyStore properties = mp_inst.query<ISetupPropertyStore>();
+                return VsSetupConfiguration::GetStringProperty(properties, L"channelId");
+            }
+
+            inline std::wstring GetChannelName() const
+            {
+                std::wstring channelName;
+
+                // channelId is in the format  <ProductName>.<MajorVersion>.<ChannelName>
+                std::wstring channelId{ GetChannelId() };
+                size_t pos = channelId.rfind(L".");
+                if (pos != std::wstring::npos)
+                {
+                    channelName.append(channelId.substr(pos + 1));
+                }
+
+                return channelName;
+            }
+
+            inline std::wstring GetProductLineVersion() const
+            {
+                ComPtrPropertyStore properties;
+                ComPtrInstanceCatalog instanceCatalog = mp_inst.query<ISetupInstanceCatalog>();

```

Comment:
`ISetupInstanceCatalog` was added late in 15.x - I forget when. This could throw, though you throw further down so maybe that's not of a concern. How does that impact overall discovery of shells, though?

Reply:
Same here. There's a non-insignificant chance this could throw so just return an empty string.

Reply:
Exceptions are caught in the BaseVisualStudioGenerator, then the profile is skipped.

Reply:
Disambiguation is required for the profile name, as the profile guid is based on the name. 

The original plan was to not support prior to 15.3, one of these methods would throw, the profile would be skipped. If we're going to support VS prior to 15.3, then we need a way to disambiguate the instances in the case the user has multiple instances of pre-15.3 installed.

So either:
1. No support for VS prior to 15.3 (I'm fine with this, given that there's an easy workaround to add the profile manually)
2. Or, we need a consistent way to disambiguate pre-15.3 instances without using ISetupPropertyStore or ISetupInstancesCatalog

Reply:
If you want to disambiguate based on something else, like the instance ID plus the shell type, that would also be acceptable. It's just a convenient shortcut that "CreateDefaultProfile" disambiguates based on the name :smile:

(You'd likely need to use the same UUIDv5 generator that CreateDefaultProfile does)

Reply:
I've decided to go with the full version for pre-15.3

Reply:
@DHowett also implemented a custom base profile creator so the guid is not based on the profile name for VS Dev profiles.